src/Core/src/Platform/iOS/MauiLabel.cs

```diff
@@ -45,10 +46,25 @@ public override void DrawText(RectangleF rect)
 			base.DrawText(rect);
 		}
 
+		public override RectangleF TextRectForBounds(RectangleF bounds, nint numberOfLines)
+		{
+			var rect = base.TextRectForBounds(TextInsets.InsetRect(bounds), numberOfLines);
+			rect.Width += (TextInsets.Left + TextInsets.Right);
+			rect.Height += (TextInsets.Top + TextInsets.Bottom);
+			return rect;
+		}
+
 		RectangleF AlignVertical(RectangleF rect)
 		{
 			var frameSize = Frame.Size;
-			var height = Lines == 1 ? Font.LineHeight : SizeThatFits(frameSize).Height;
+			var textSize = new NSString(Text).GetBoundingRect(
+				rect.Size,
+				NSStringDrawingOptions.UsesLineFragmentOrigin | NSStringDrawingOptions.UsesFontLeading,
+				new UIStringAttributes { Font = Font },
+				null
+			).Size;
+
+			var height = Lines == 1 ? Font.LineHeight : textSize.Height;
 
 			if (height < frameSize.Height)
```

Comment:
should this be updated to compare against `textSize` as well?
Is there any purpose in still using `frameSize`? 

(not tested and not reviewing, just reading the code and asking here)

Reply:
Hi, @rachelkang 
Thank you for sparing your valuable time.

Before stating the answer to the question, I found a problem with the implementation I showed.
It is the lack of consideration for character decoration and spacing. I will reconsider.

The reason I used textSize instead of frameSize is that using frameSize rounds the height of 
the Label to the height of parent. 

For example, if the parent height is 160 and the label padding is 20 above and below, 
the maximum label height is 120. If the height of one line of Label is 36, the maximum number of lines 
that can be displayed in Label is 3. If 4 lines of text are specified for the label, the height obtained 
by the SizeThatFits method is 184, which is the height of 4 lines (36 * 4 = 144) plus the padding 
top and bottom (20 + 20 = 40). , but it will be rounded to the height of the parent.
In other words, the height obtained by calling the SizeThatFits method is 160.

Subtracting the top and bottom of the padding from that, 120 is the height of the label when 
4 lines are set, but since the number of lines that can be displayed is 3 lines, the height must be 108.
So even if you call the SizeThatFits method and subtract the padding top and bottom, you won't 
get 108, which is 3 lines high.

So I came up with a way to make use of the GetBoundingRect method. This is because it does 
not round due to the size of the parent. If the label is only tall enough to display a maximum of 
3 lines, the return value of the GetBoundingRect method will only return the height of 3 lines.

Since there is no gap, the label will be displayed at the intended position when you specify Start or 
End for VerticalOptions.

It looked fine, but as I mentioned earlier, it lacks consideration for letter decoration and spacing.
I also need to verify that there is no difference between the size obtained by GetBoundingRect 
and the actual size of the displayed label, which is not ideal.

Therefore, we will separately consider how to use only the SizeThatFits method without relying 
on the GetBoundingRect method. As soon as it is uploaded again, we will announce it here.

Reply:
I thought of a method that does not use the size calculation by the GetBoundingRect method.
This PR solves the issue by using the number of lines currently displayed in the label and the 
maximum number of lines that can be displayed in the display area.

Add the following class variable to store the number of lines displayed for the current label.

    nfloat _currentRowCount = 0;

In the SizeThatFits method, I store the number of lines that were displayed in the label 
before it was rounded to the size of the parent in the class variable from earlier.

    public override SizeF SizeThatFits(SizeF size)
    {
        var requestedSize = base.SizeThatFits(size);

        _currentRowCount = NMath.Floor((requestedSize.Height - (TextInsets.Top + TextInsets.Bottom)) / Font.LineHeight);

        // Let's be sure the label is not larger than the container
        return new Size()
        {
            Width = nfloat.Min(requestedSize.Width, size.Width),
            Height = nfloat.Min(requestedSize.Height, size.Height),
        };
    }

Within the AlignVertical method, calculate the maximum number of lines that can be displayed 
from the display area. If the number of lines currently displayed exceeds the maximum number 
of lines that can be displayed, the height of the label is rounded by the height calculated 
by the maximum number of lines that can be displayed.

    RectangleF AlignVertical(RectangleF rect)
    {
        var frameSize = Frame.Size;
        var height = Lines == 1 ? Font.LineHeight : SizeThatFits(frameSize).Height;

        if (Lines != 1)
        {
            nfloat maximumRowCount = NMath.Floor((Frame.Size.Height - (TextInsets.Top + TextInsets.Bottom)) / Font.LineHeight);

            if (_currentRowCount > maximumRowCount)
            {
                height = maximumRowCount * Font.LineHeight;
            }
            else
            {
                height -= (TextInsets.Top + TextInsets.Bottom);
            }
        }

        if (height < frameSize.Height)
        {
            if (_verticalAlignment == UIControlContentVerticalAlignment.Top)
            {
                rect.Height = height;
            }
            else if (_verticalAlignment == UIControlContentVerticalAlignment.Bottom)
            {
                rect = new RectangleF(rect.X, rect.Bottom - height, rect.Width, height);
            }
        }

        return rect;
    }

This gives the same result as using the GetBoundingRect method. This correspondence 
is valid even if CharacterSpacing is specified.

I tested it with the layout below.

    <Grid RowDefinitions="44,160,*" ColumnDefinitions="*,*" RowSpacing="5">
        <Entry x:Name="txtTest" Grid.Row="0" TextChanged="txtTest_TextChanged" />
        <Grid Grid.Row="1" Grid.ColumnSpan="2" BackgroundColor="Red" Padding="0,0,0,0" />
        <Label x:Name="lblTest" Grid.Row="1" LineBreakMode="WordWrap" FontSize="30" Padding="20,20,20,20" TextColor="White" VerticalOptions="Start" VerticalTextAlignment="Start" BackgroundColor="Blue" HorizontalOptions="Fill" />
    </Grid>

Below are the verification results.

`CharacterSpacing="5"`

https://github.com/dotnet/maui/assets/125236133/90dd3b91-566e-4127-9fe5-4ef3be22f2e8

`FontAttributes="Bold"`

https://github.com/dotnet/maui/assets/125236133/758ab227-029a-47a3-b878-2d3862079a77

`FontAttributes="Italic"`

https://github.com/dotnet/maui/assets/125236133/85ebc53a-4328-45e0-862c-ccbc7cc15ea9

Other verification results are the same as those uploaded before the change.

It would be nice to be able to write test cases, but I have to judge by appearance, so what should I do?
Sorry for my lack of power.

Reply:
Changed not to apply size limit only when calling SizeThatFits method in DrawText method.
As a result, when the text that exceeds the height of the display area is set to the Label, 
the text will be displayed over the padding, but it will be almost the same movement as Android.

    RectangleF AlignVertical(RectangleF rect)
    {
        var frameSize = Frame.Size;

        _isLimitSize = false;
        var height = Lines == 1 ? Font.LineHeight : SizeThatFits(frameSize).Height;
        _isLimitSize = true;

        if (Lines != 1)
        {
            height -= (TextInsets.Top + TextInsets.Bottom);
        }

        if (_verticalAlignment == UIControlContentVerticalAlignment.Top)
        {
            rect.Height = height;
        }
        else if (_verticalAlignment == UIControlContentVerticalAlignment.Bottom)
        {
            rect = new RectangleF(rect.X, rect.Bottom - height, rect.Width, height);
        }

        return rect;
    }

    public override SizeF SizeThatFits(SizeF size)
    {
        var requestedSize = base.SizeThatFits(size);

        if (_isLimitSize)
        {
            // Let's be sure the label is not larger than the container
            return new Size()
            {
                Width = nfloat.Min(requestedSize.Width, size.Width),
                Height = nfloat.Min(requestedSize.Height, size.Height),
            };
        }
        else
        {
            return requestedSize;
        }
    }

I removed the text that covered the padding area by clipping it using a mask.

    public override void DrawText(RectangleF rect)
    {
        rect = TextInsets.InsetRect(rect);
        var clipRect = new RectangleF(rect.X, rect.Y, rect.Width, rect.Height);

        if (_verticalAlignment != UIControlContentVerticalAlignment.Center
            && _verticalAlignment != UIControlContentVerticalAlignment.Fill)
        {
            rect = AlignVertical(rect);
        }

        base.DrawText(rect);

        var maskLayer = new CAShapeLayer() { Frame = Bounds, Path = UIBezierPath.FromRect(clipRect).CGPath };
        maskLayer.MasksToBounds = true;
        Layer.Mask = maskLayer;
    }

This method works as intended without using NSString's GetBoundingRect method or c
alculating the height of one line.

With this fix, it behaves almost identically to Android.
A merge of PR #16223 is required to apply this PR.

I can't upload today, so I'll shoot a verification video tomorrow and upload it.
Before I write the tests, I'd like your opinion on whether this fix is ​​what you'd expect.

Reply:
Below are the verification results.

I verified it by changing the Label property in the layout below.

    <Grid RowDefinitions="44,160,*" ColumnDefinitions="*,*" RowSpacing="5">
        <Entry x:Name="txtTest" Grid.Row="0" Grid.ColumnSpan="2" TextChanged="txtTest_TextChanged"  />
        <Grid Grid.Row="1" Grid.Column="0" BackgroundColor="Red" Margin="20,20,20,20" />
        <Label x:Name="lblTest" Grid.Row="1" Grid.Column="0" LineBreakMode="WordWrap" FontSize="30" Padding="20,20,20,20" TextColor="White" VerticalOptions="Start" VeticalTextAlignment="Start" />
    </Grid>

    private void txtTest_TextChanged(object sender, TextChangedEventArgs e)
    {
        lblTest.Text = txtTest.Text;
    }

`VerticalOptions="Start" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/e3f295bf-7796-4b08-9aab-914af2cc7e3e

`VerticalOptions="Start" VeticalTextAlignment="Center"`

https://github.com/dotnet/maui/assets/125236133/79dd5b52-1f87-43b7-9c0c-91508b6ceb90

`VerticalOptions="Start" VeticalTextAlignment="End"`

https://github.com/dotnet/maui/assets/125236133/6514eab1-2975-4578-aef6-52a08ab17a10

`VerticalOptions="Center" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/8fedba1c-06c5-4e03-b295-f20e17cc23cb

`VerticalOptions="Center" VeticalTextAlignment="Center"`

https://github.com/dotnet/maui/assets/125236133/09634000-1d83-4e2a-870d-4574bdafbb81

`VerticalOptions="Center" VeticalTextAlignment="End"`

https://github.com/dotnet/maui/assets/125236133/ec720f70-309f-4b47-8c61-b2f52f8048bc

`VerticalOptions="End" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/81551d05-c20b-405c-b96d-c98d7e8e2df4

`VerticalOptions="End" VeticalTextAlignment="Center"`

https://github.com/dotnet/maui/assets/125236133/7b827281-7a56-411f-ad40-97e19991bf2f

`VerticalOptions="End" VeticalTextAlignment="End"`

https://github.com/dotnet/maui/assets/125236133/2a7caadb-65aa-4a33-a9ea-de9ca4fbacf9

`VerticalOptions="Fill" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/f812394e-02ee-41f0-8980-be9fdaff1681

`VerticalOptions="Fill" VeticalTextAlignment="Center"`

https://github.com/dotnet/maui/assets/125236133/e3b64083-cdee-43ef-a8b6-61db72fa9c1e

`VerticalOptions="Fill" VeticalTextAlignment="End"`

https://github.com/dotnet/maui/assets/125236133/08a7b429-2410-42e6-97d2-4ce6cef34df4

`HorizontalOptions="Start" VerticalOptions="Start" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/e5af7c61-60ae-43a6-bb69-94856a113665

`HorizontalOptions="Center" VerticalOptions="Start" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/0b4202f4-1f6e-41d3-bd24-8910a8d90a46

`HorizontalOptions="End" VerticalOptions="Start" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/da1fc4c1-b28d-4701-a795-e4afd0a679e1

`LineBreakMode="NoWrap" VerticalOptions="Start" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/5c2befec-084f-416c-9ffb-c9977141bccb

`LineBreakMode="CharacterWrap" VerticalOptions="Start" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/4ac37330-c1a0-4bc8-911c-c439f8ff4476

`LineBreakMode="HeadTruncation" VerticalOptions="Start" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/3535c5d3-1f0c-4865-87bc-51f04e45009b

`LineBreakMode="MiddleTruncation" VerticalOptions="Start" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/f578d3da-5089-4bbf-a8e0-57ede5c487f2

`LineBreakMode="TailTruncation" VerticalOptions="Start" VeticalTextAlignment="Start"`

https://github.com/dotnet/maui/assets/125236133/ee40b364-25a9-4a77-9ccf-772ae7aca71a

`CharacterSpacing="5"`

https://github.com/dotnet/maui/assets/125236133/b6067181-4dc5-47eb-ac9c-f8e15dbf1fea

`LineHeight="1.2"`

https://github.com/dotnet/maui/assets/125236133/56ee22e9-1b43-4a1d-81af-8001067205f9

`FontAttributes="Bold"`

https://github.com/dotnet/maui/assets/125236133/f206f763-e008-43c5-9367-7ed473409f03

`FontAttributes="Italic"`

https://github.com/dotnet/maui/assets/125236133/9085f046-106b-4a3e-afcf-431249226bb4

I set the characters entered in Entry in the Span of the following layout and verified it.

    <Grid RowDefinitions="44,160,*" ColumnDefinitions="*,*" RowSpacing="5">
        <Entry x:Name="txtTest" Grid.Row="0" Grid.ColumnSpan="2" TextChanged="txtTest_TextChanged"  />
        <Grid Grid.Row="1" Grid.Column="0" BackgroundColor="Red" Margin="20,20,20,20" />
        <Label x:Name="lblTest" Grid.Row="1" Grid.Column="0" LineBreakMode="WordWrap" FontSize="30" Padding="20,20,20,20" TextColor="Black" VerticalOptions="Start" VerticalTextAlignment="Start" HorizontalOptions="Start">
            <Label.FormattedText>
                <FormattedString>
                    <Span x:Name="spnTest1" FontSize="30" TextColor="White" BackgroundColor="Blue" />
                    <Span x:Name="spnTest2" FontSize="20" TextColor="White" BackgroundColor="Blue" />
                </FormattedString>
            </Label.FormattedText>
        </Label>
    </Grid>

    private void txtTest_TextChanged(object sender, TextChangedEventArgs e)
    {
        spnTest1.Text = txtTest.Text;
        spnTest2.Text = txtTest.Text;
    }

https://github.com/dotnet/maui/assets/125236133/80d258a1-41e2-4c8e-965d-e0d79b893351

I've verified by inserting HTML tags into the characters entered in the entry in the layout below.

    <Grid RowDefinitions="44,160,*" ColumnDefinitions="*,*" RowSpacing="5">
        <Entry x:Name="txtTest" Grid.Row="0" Grid.ColumnSpan="2" TextChanged="txtTest_TextChanged"  />
        <Grid Grid.Row="1" Grid.Column="0" BackgroundColor="Red" Margin="20,20,20,20" />
        <Label x:Name="lblTest" Grid.Row="1" Grid.Column="0" LineBreakMode="WordWrap" Padding="20,20,20,20" VerticalOptions="Start" VerticalTextAlignment="Start" HorizontalOptions="Start" TextType="Html" />
    </Grid>

    private void txtTest_TextChanged(object sender, TextChangedEventArgs e)
    {
        lblTest.Text = "<font size=\"25\" color=\"white\">" + txtTest.Text + "</font>";
    }

https://github.com/dotnet/maui/assets/125236133/abc59aee-8964-4d84-9df3-5981a46fc6ac

I have verified by placing a label on the ScrollView.

    <Grid RowDefinitions="44,160,*" ColumnDefinitions="*,*" RowSpacing="5">
        <Entry x:Name="txtTest" Grid.Row="0" Grid.ColumnSpan="2" TextChanged="txtTest_TextChanged"  />
        <Grid Grid.Row="1" Grid.Column="0" BackgroundColor="Red" Margin="20,20,20,20" />
        <ScrollView Grid.Row="1" Orientation="Vertical">
            <Label x:Name="lblTest" LineBreakMode="WordWrap" FontSize="30" Padding="20,20,20,20" TextColor="White" VerticalOptions="Start" VerticalTextAlignment="Start" BackgroundColor="Blue" />
        </ScrollView>
    </Grid>

https://github.com/dotnet/maui/assets/125236133/c6cde762-8bd5-4bff-8171-b4f7c621491e

I verified with the Grid's RowDefinitions set to Auto.

    <Grid RowDefinitions="44,Auto,*" ColumnDefinitions="*,*" RowSpacing="5">
        <Entry x:Name="txtTest" Grid.Row="0" Grid.ColumnSpan="2" TextChanged="txtTest_TextChanged"  />
        <Grid Grid.Row="1" Grid.Column="0" BackgroundColor="Red" Margin="20,20,20,20" />
        <Label x:Name="lblTest" Grid.Row="1" Grid.Column="0" LineBreakMode="WordWrap" FontSize="30" Padding="20,20,20,20" TextColor="White" VerticalOptions="Start" VeticalTextAlignment="Start" />
    </Grid>

https://github.com/dotnet/maui/assets/125236133/3d80d02e-1dfa-4bcc-b2ad-bbd070d0716d

Could you please do a review?

The expected value is the size of the label's display area when the text is set, but I don't know how to calculate that size.
I'm confused as to how to calculate it. I understand how to write test code, but it is difficult to calculate the size including wrapping.
Any good ideas? Thank you.